## AIND - Planning Lab Analysis


This notebook intends to make an analysis in order to fully complete AIND Planning Lab.

### Part 1 - Planning problems:

After coding `my_air_cargo_problems`, it be showed results running different search methods.




### First problem: air_cargo_p1

- Search methods: breadth_first_search and breadth_first_tree_search

In [4]:
%run run_search.py -p 1 -s 1


Solving Air Cargo Problem 1 using breadth_first_search...

Expansions   Goal Tests   New Nodes
    43          56         180    

Plan length: 6  Time elapsed in seconds: 0.058320423937402666
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Fly(P2, JFK, SFO)
Unload(C2, P2, SFO)
Fly(P1, SFO, JFK)
Unload(C1, P1, JFK)



In [5]:
%run run_search.py -p 1 -s 2


Solving Air Cargo Problem 1 using breadth_first_tree_search...

Expansions   Goal Tests   New Nodes
   1458        1459        5960   

Plan length: 6  Time elapsed in seconds: 1.5204992319922894
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Fly(P2, JFK, SFO)
Unload(C2, P2, SFO)
Fly(P1, SFO, JFK)
Unload(C1, P1, JFK)



- Search methods: **depth_first_graph_search** and **depth_limited_search**

In [6]:
%run run_search.py -p 1 -s 3


Solving Air Cargo Problem 1 using depth_first_graph_search...

Expansions   Goal Tests   New Nodes
    21          22          84    

Plan length: 20  Time elapsed in seconds: 0.022511297021992505
Fly(P1, SFO, JFK)
Fly(P2, JFK, SFO)
Load(C2, P1, JFK)
Fly(P1, JFK, SFO)
Fly(P2, SFO, JFK)
Unload(C2, P1, SFO)
Fly(P1, SFO, JFK)
Fly(P2, JFK, SFO)
Load(C2, P2, SFO)
Fly(P1, JFK, SFO)
Load(C1, P2, SFO)
Fly(P2, SFO, JFK)
Fly(P1, SFO, JFK)
Unload(C2, P2, JFK)
Unload(C1, P2, JFK)
Fly(P2, JFK, SFO)
Load(C2, P1, JFK)
Fly(P1, JFK, SFO)
Fly(P2, SFO, JFK)
Unload(C2, P1, SFO)



In [7]:
%run run_search.py -p 1 -s 4


Solving Air Cargo Problem 1 using depth_limited_search...

Expansions   Goal Tests   New Nodes
   101         271         414    

Plan length: 50  Time elapsed in seconds: 0.15404366701841354
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1, P1, SFO)
Load(C1, P1, SFO)
Unload(C1,

- Search method: **uniform_cost_search**

In [9]:
%run run_search.py -p 1 -s 5


Solving Air Cargo Problem 1 using uniform_cost_search...

Expansions   Goal Tests   New Nodes
    55          57         224    

Plan length: 6  Time elapsed in seconds: 0.07320064795203507
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Fly(P1, SFO, JFK)
Fly(P2, JFK, SFO)
Unload(C1, P1, JFK)
Unload(C2, P2, SFO)



### First problem: air_cargo_p2

- Search methods: breadth_first_search and breadth_first_tree_search

In [10]:
%run run_search.py -p 2 -s 1


Solving Air Cargo Problem 2 using breadth_first_search...

Expansions   Goal Tests   New Nodes
   3343        4609       30509   

Plan length: 9  Time elapsed in seconds: 19.281205381965265
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Load(C3, P3, ATL)
Fly(P2, JFK, SFO)
Unload(C2, P2, SFO)
Fly(P1, SFO, JFK)
Unload(C1, P1, JFK)
Fly(P3, ATL, SFO)
Unload(C3, P3, SFO)



**Result from breadth_first_tree_search**: It took more than 3 minutes

- Search methods: **depth_first_graph_search** and **depth_limited_search**

In [13]:
%run run_search.py -p 2 -s 3


Solving Air Cargo Problem 2 using depth_first_graph_search...

Expansions   Goal Tests   New Nodes
   624         625         5602   

Plan length: 619  Time elapsed in seconds: 4.86183173104655
Fly(P3, ATL, SFO)
Fly(P1, SFO, ATL)
Fly(P3, SFO, JFK)
Fly(P1, ATL, JFK)
Fly(P2, JFK, ATL)
Fly(P3, JFK, ATL)
Fly(P2, ATL, SFO)
Fly(P3, ATL, SFO)
Load(C2, P1, JFK)
Fly(P2, SFO, ATL)
Fly(P1, JFK, ATL)
Fly(P2, ATL, JFK)
Fly(P1, ATL, SFO)
Fly(P3, SFO, ATL)
Fly(P1, SFO, JFK)
Load(C3, P3, ATL)
Fly(P3, ATL, SFO)
Fly(P2, JFK, ATL)
Fly(P3, SFO, JFK)
Fly(P2, ATL, SFO)
Fly(P1, JFK, ATL)
Fly(P2, SFO, JFK)
Fly(P1, ATL, SFO)
Unload(C3, P3, JFK)
Fly(P1, SFO, JFK)
Fly(P3, JFK, ATL)
Fly(P2, JFK, ATL)
Fly(P3, ATL, SFO)
Fly(P2, ATL, SFO)
Fly(P1, JFK, ATL)
Fly(P3, SFO, ATL)
Fly(P1, ATL, SFO)
Unload(C2, P1, SFO)
Fly(P3, ATL, SFO)
Fly(P2, SFO, ATL)
Fly(P1, SFO, ATL)
Fly(P2, ATL, JFK)
Fly(P1, ATL, JFK)
Fly(P3, SFO, ATL)
Fly(P2, JFK, ATL)
Fly(P3, ATL, JFK)
Fly(P2, ATL, SFO)
Fly(P1, JFK, ATL)
Load(C3, P3, JFK)
Fly(P1, 

**Result from depth_limited_search**: It took more than 3 minutes

### First problem: air_cargo_p3

- Search methods: breadth_first_search and breadth_first_tree_search

In [1]:
%run run_search.py -p 3 -s 1


Solving Air Cargo Problem 3 using breadth_first_search...

Expansions   Goal Tests   New Nodes
  14663       18098       129631  

Plan length: 12  Time elapsed in seconds: 132.7554389850702
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Fly(P2, JFK, ORD)
Load(C4, P2, ORD)
Fly(P1, SFO, ATL)
Load(C3, P1, ATL)
Fly(P1, ATL, JFK)
Unload(C1, P1, JFK)
Unload(C3, P1, JFK)
Fly(P2, ORD, SFO)
Unload(C2, P2, SFO)
Unload(C4, P2, SFO)



**Result from breadth_first_tree_search**: It took more than 5 minutes

- Search methods: **depth_first_graph_search** and **depth_limited_search**

In [3]:
%run run_search.py -p 3 -s 3


Solving Air Cargo Problem 3 using depth_first_graph_search...

Expansions   Goal Tests   New Nodes
   408         409         3364   

Plan length: 392  Time elapsed in seconds: 2.3902184600010514
Fly(P1, SFO, ORD)
Fly(P2, JFK, ORD)
Fly(P1, ORD, ATL)
Fly(P2, ORD, ATL)
Fly(P1, ATL, JFK)
Fly(P2, ATL, SFO)
Load(C2, P1, JFK)
Fly(P2, SFO, ORD)
Fly(P1, JFK, ORD)
Fly(P2, ORD, ATL)
Fly(P1, ORD, ATL)
Fly(P2, ATL, JFK)
Fly(P1, ATL, SFO)
Unload(C2, P1, SFO)
Fly(P1, SFO, ORD)
Fly(P2, JFK, ORD)
Fly(P1, ORD, ATL)
Fly(P2, ORD, ATL)
Fly(P1, ATL, JFK)
Fly(P2, ATL, SFO)
Load(C2, P2, SFO)
Fly(P2, SFO, ORD)
Fly(P1, JFK, ORD)
Fly(P2, ORD, ATL)
Fly(P1, ORD, ATL)
Fly(P2, ATL, JFK)
Load(C3, P1, ATL)
Fly(P1, ATL, ORD)
Fly(P2, JFK, ORD)
Fly(P1, ORD, SFO)
Fly(P2, ORD, ATL)
Fly(P1, SFO, JFK)
Fly(P2, ATL, SFO)
Unload(C3, P1, JFK)
Fly(P2, SFO, ORD)
Fly(P1, JFK, ORD)
Fly(P2, ORD, ATL)
Fly(P1, ORD, ATL)
Fly(P2, ATL, JFK)
Fly(P1, ATL, SFO)
Unload(C2, P2, JFK)
Fly(P1, SFO, ORD)
Fly(P2, JFK, ORD)
Fly(P1, ORD, ATL)
Fly(

- Search method: **uniform_cost_search**

In [4]:
%run run_search.py -p 3 -s 5


Solving Air Cargo Problem 3 using uniform_cost_search...

Expansions   Goal Tests   New Nodes
  18151       18153       159038  

Plan length: 12  Time elapsed in seconds: 445.7282589379465
Load(C1, P1, SFO)
Load(C2, P2, JFK)
Fly(P1, SFO, ATL)
Load(C3, P1, ATL)
Fly(P2, JFK, ORD)
Load(C4, P2, ORD)
Fly(P2, ORD, SFO)
Fly(P1, ATL, JFK)
Unload(C1, P1, JFK)
Unload(C2, P2, SFO)
Unload(C3, P1, JFK)
Unload(C4, P2, SFO)



### Summary from results

#### Air Cargo Problem 1





|   |Expansions   |  Goal Tests |  New Nodes  | Plan length |Time elapsed in seconds |
|---|---          |---          |---          |---          |---                    |
|breadth_first_search   | 43  | 56  |180   | 6 | 0.0583
| breadth_first_tree_search  |  1458 |  1459 | 5960  | 6 | 1.520
|depth_first_graph_search   | 21  | 22  | 84  | 20 | 0.022
|  depth_limited_search | 101  |  271 | 414  | 50 | 0.154
| uniform_cost_search | 55   | 57 | 224 | 6 | 0.073


#### Air Cargo Problem 2

|   |Expansions   |  Goal Tests |  New Nodes  | Plan length |Time elapsed in seconds |
|---|---          |---          |---          |---          |---                    |
|breadth_first_search   | 3343  | 4609  |30509   | 9 | 19.28120
| breadth_first_tree_search  |  It tooks more than 3 minutes | 
|depth_first_graph_search   | 624  | 625  | 5602  | 619 | 4.861831
|  depth_limited_search | It tooks more than 3 minutes| 


#### Air Cargo Problem 3

|   |Expansions   |  Goal Tests |  New Nodes  | Plan length |Time elapsed in seconds |
|---|---          |---          |---          |---          |---                    |
|breadth_first_search   | 14663  | 18098  |129631   | 12 | 132.755
| breadth_first_tree_search  |  It tooks more than 5 minutes |
|depth_first_graph_search   | 408  | 409  | 3364  | 392 | 2.39021
|  depth_limited_search | It tooks more than 3 minutes |
|uniform_cost_search   | 18151  | 18153  | 159038  | 12 | 445.728

